Geoserver setup

f-PLT opened the pull request:
* Add Geoserver workspace creation
* Add Geoserver datastore creation
* Add shapefile publishing

Specifics points to look at:

* The tests are failing right now at the `Install dependencies` (prior to this PR), looks like it's hanging on `docformatter-0.5.6.tar.gz` . I haven't investigated yet, in case it's a known problem.

* Right now, tests will fail on Github because there is no Geoserver setup for testing. My question is, do we want Github actions to run Geoserver like it does with Mongodb? Geoserver is a somewhat big image (close to 1GB) and I'm not sure if it would be worth it to test at this level vs Birdhouse level.

* I implemented shapefile publishing using the structures and mechanisms of the Catalog service, as it seemed the most efficient way of doing it, but I'm open to other implementations if you have ideas/objections.

* I have not created tests for shapefile publishing yet as they are a bit more involved (need to have a valid shapefile for the test) and there's no real substitute to a visual inspection to make sure the data is well configured and displays well as a WFS. I could relatively quickly test if the requests runs well and returns the right code but the value of that test is limited.

* Again for the shapefile publishing, I'm anticipating a "user-side" problem. Since shapefile is a multi file format, if all parts are not copied/moved to the user's `shapefile_datastore` folder, there there could be some problems (especially if the `.proj` file is added later). A simple way would be to make the job wait for a few seconds after it registers a `.shp` file. The other, more involved, would be to check for the required extensions and only start the publishing process after all the required extensions are present.

f-PLT committed:
Add Geoserver do docker-compose-example

f-PLT committed:
Update config.example.yml

Set url, add admin_user and admin_password for geoserver service

f-PLT committed:
Add admin and password parameters for services

f-PLT committed:
Implement Geoserver workspace creation

f-PLT committed:
Implement Geoserver datastore creation

f-PLT committed:
Cleanup geoserver service error management

f-PLT committed:
Implement Geoserver workspace deletion

f-PLT committed:
Harmonize Geoserver logs with other services

f-PLT committed:
Enable celery worker tasks for Geoserver service

f-PLT committed:
Update type hints Geoserver service

f-PLT committed:
Update existing geoserver tests

f-PLT committed:
Refactor Geoserver service

Split Geoserver requests from the service tasks  for better
responsibility encapsulation and easier request testing.

f-PLT committed:
Reorder args in `_configure_datastore_request()`

f-PLT committed:
Add tests for Geoserver requests

f-PLT committed:
Remove todo in test_request_task.py : test_geoserver

f-PLT committed:
Add missing quotations for container ports

f-PLT committed:
Add geoserver pytest fixture

f-PLT committed:
Add test target for geoserver in makefile

f-PLT committed:
Add publish_shapefile feature to Geoserver service

f-PLT committed:
Add publish_shapefile to catalog's on_create

f-PLT committed:
Implement remove_shapefile in Geoserver service

f-PLT committed:
Change to double quotes in geoserver.py

f-PLT committed:
Fix rebase mistakes

f-PLT committed:
Update test_geoserver.py to use ServiceFactory

However, now the Env variables need to be set for the
tests to succeed.

f-PLT committed:
Cleanup test_geoserver.py

f-PLT committed:
Temporary fix for TestSyncPermissions tests

f-PLT committed:
Workaround for Geoserver tests side effects

The use of ServiceFactory in `test_geoserver.py` caused
other tests to fail. Instead, falling back to building separate
Geoserver instance.

f-PLT committed:
Update AUTHORS and CHANGES files

f-PLT committed:
Update geoserver service docstring

f-PLT committed:
Add Geoserver user_deleted test

dbyrns commented on cowbird/services/service.py:
```diff
@@ -12,11 +12,15 @@
     SERVICE_PRIORITY_PARAM,
     SERVICE_URL_PARAM,
     SERVICE_WORKSPACE_DIR_PARAM,
+    SERVICE_ADMIN_USER,
```
My bad here (not enough documentation), but those parameters were intended for "shared" parameters handled by the base service class. For service specific parameters like geoserver user/pwd you could fetch them directly here : https://github.com/Ouranosinc/cowbird/blob/master/cowbird/services/impl/geoserver.py#L23 from kwargs without declaring them anywhere else.

dbyrns commented on docker/docker-compose.example.yml:
```diff
@@ -48,3 +48,41 @@ services:
+    environment:
+      - GEOSERVER_DATA_DIR=/opt/footprints_dir
+      - GEOWEBCACHE_CACHE_DIR=/opt/geoserver/data_dir/gwc
+      - GEOSERVER_ADMIN_USER=admin
```
Use the same env var here for user/pwd.

dbyrns commented on docker/docker-compose.example.yml:
```diff
@@ -48,3 +48,41 @@ services:
+      - "8765:8080"
+    restart: on-failure
+    volumes:
+      - /user_workspaces:/user_workspaces
```
See cowbird here : https://github.com/Ouranosinc/cowbird/blob/master/docker/docker-compose.override.example.yml#L20
A local directory was used to avoid requiring that directory on the dev machine. (It is even commited (https://github.com/Ouranosinc/cowbird/tree/master/docker/user_workspaces), but sadly I saw that some artefacts got commited there, a .gitignore should be added to avoid comitting user directory)

fmigneault replied:
I think it would be better to set host path on `/tmp/user_workspaces` to avoid dumping directories all over the place. User should provide a specific override (or edit file) if it wants to persist data from it.

f-PLT replied:
I had a talk with @dbyrns on this. I ended up replacing everything with environment variables, as discussed elsewhere, but in https://github.com/Ouranosinc/cowbird/blob/geoserver-setup/docker/.env.example I did set it to the local directory mentioned by him as we both found it was easier for manual testing.

fmigneault replied:
I agree with using variables to simplify setup without the need to edit any YML file. 
I still disagree with using (default) local relative paths though. 
It should be a tmp dir, which both locally for testing or in CI can be overridden with the variable.
This is to avoid mistake commits such as https://github.com/Ouranosinc/cowbird/tree/master/docker/user_workspaces

dbyrns commented on cowbird/services/impl/geoserver.py:
```diff
@@ -50,42 +59,335 @@ def permission_deleted(self, permission):
+                                                  filename=filename)
+        response_code = response.status_code
+        if response_code == 200:
+            LOGGER.info("Shapefile [%s] has been successfully publish by Geoserver.", filename)
```
copy/paste error

dbyrns commented on cowbird/services/impl/geoserver.py:
```diff
@@ -50,42 +59,335 @@ def permission_deleted(self, permission):
+        elif response_code == 500:
+            LOGGER.error(response.text)
+        else:
+            LOGGER.error("There was an error publishing the following shapefile: [%s]", filename)
```
copy/paste error

dbyrns commented on cowbird/services/impl/geoserver.py:
```diff
@@ -50,42 +59,335 @@ def permission_deleted(self, permission):
+    def _create_datastore_dir(datastore_path):
+        # type (str) -> None
+        try:
+            os.mkdir(datastore_path)
```
Could be a great place to start monitoring for new shapefiles

dbyrns commented on cowbird/services/impl/catalog.py:
```diff
@@ -62,6 +62,16 @@ def on_created(self, filename):
         :param filename: Relative filename of a new file
         """
         LOGGER.info("The following file [%s] has just been created", filename)
+        # TODO This is a first naive implementation. Needs to be revised and improved.
```
This implementation should be in `impl/geoserver.py`
All you need to do is : 
-  Implement `FSMonitor` from `Geoserver` like here : https://github.com/Ouranosinc/cowbird/blob/master/cowbird/services/impl/catalog.py#L15
-  Register/unregister the `Geoserver` class like here : https://github.com/Ouranosinc/cowbird/blob/master/cowbird/services/impl/catalog.py#L38, https://github.com/Ouranosinc/cowbird/blob/master/cowbird/services/impl/catalog.py#L42
- Implement on_created, on_deleted and on_modified. Basically, moving the code from here.

dbyrns reviewed (changes requested):
The essential is already there. I added some precisions on where stuff should be put, but nothing hard to fix.

dbyrns commented:
> The tests are failing right now at the Install dependencies (prior to this PR), looks like it's hanging on docformatter-0.5.6.tar.gz . I haven't investigated yet, in case it's a known problem.

I will let @fmigneault gives its input on this problem.

> Right now, tests will fail on Github because there is no Geoserver setup for testing. My question is, do we want Github actions to run Geoserver like it does with Mongodb? Geoserver is a somewhat big image (close to 1GB) and I'm not sure if it would be worth it to test at this level vs Birdhouse level.

I don't think that we want to go there. I really appreciate the tests you added, and they could be helpful while working specifically on geoserver on the dev side, but if we could disable them for the CI it would be perfect. I guess that these kind of tests would be better on a notebook for a full stack test.

> I implemented shapefile publishing using the structures and mechanisms of the Catalog service, as it seemed the most efficient way of doing it, but I'm open to other implementations if you have ideas/objections.

I added specific comments on that topic. The current monitoring implementation is, I agree, not efficient, but the internal implementation could monitor paths once and notify multiple observers. In my opinion, that would be a better optimisation than piggybacking the `Catalog` implementation with `Geoserver` stuff.

> I have not created tests for shapefile publishing yet as they are a bit more involved (need to have a valid shapefile for the test) and there's no real substitute to a visual inspection to make sure the data is well configured and displays well as a WFS. I could relatively quickly test if the requests runs well and returns the right code but the value of that test is limited.

Given the previous answer for tests, I'm ok to wait until we have notebooks to refine the tests.

> Again for the shapefile publishing, I'm anticipating a "user-side" problem. Since shapefile is a multi file format, if all parts are not copied/moved to the user's shapefile_datastore folder, there there could be some problems (especially if the .proj file is added later). A simple way would be to make the job wait for a few seconds after it registers a .shp file. The other, more involved, would be to check for the required extensions and only start the publishing process after all the required extensions are present.

Could we first check the integrity of a shapefile and stop the job if the shapefile is not complete. Any new/modified file should trigger a new event. At some point an event should yield a valid shapefile that can be published?

fmigneault commented:
@f-PLT @dbyrns 
> > The tests are failing right now at the Install dependencies (prior to this PR), looks like it's hanging on docformatter-0.5.6.tar.gz . I haven't investigated yet, in case it's a known problem.
> 
> I will let @fmigneault gives its input on this problem.

I will need an output log or something to analyze the issue.
Package `docformatter` has not had any new version since Dec 2020, so I doubt it is (directly) the cause.

fmigneault commented:
@f-PLT @dbyrns 
> Right now, tests will fail on Github because there is no Geoserver setup for testing. My question is, do we want Github actions to run Geoserver like it does with Mongodb? Geoserver is a somewhat big image (close to 1GB) and I'm not sure if it would be worth it to test at this level vs Birdhouse level.

I suggest using package `responses` to mock sending requests to a "Geoserver" endpoint to validate basic operations. 
You only need to register the expected URLs and the response contents to return instead.
Example below where I mock a WPS-1 server to send `GetCapabilities` and `DescribeProcess` requests: 
https://github.com/crim-ca/weaver/blob/master/tests/utils.py#L436-L515


For bigger tests, that are closer to the resulting stack with interactions between Magpie/Cowbird/Geoserver, I would use notebooks as suggested by @dbyrns.

fmigneault commented on tests/test_geoserver.py:
```diff
@@ -0,0 +1,109 @@
+    return geoserver_settings
+
+
+GEOSERVER_SETTINGS = get_geoserver_settings()
```
Should be placed in the test class, like below `workspaces`.

f-PLT commented:
@dbyrns 

I agree with pretty much everything in your comments and will start working on them later today.

> I don't think that we want to go there. I really appreciate the tests you added, and they could be helpful while working specifically on geoserver on the dev side, but if we could disable them for the CI it would be perfect. I guess that these kind of tests would be better on a notebook for a full stack test.

That was my feeling. I'll created a new target in the Makefile to run all tests but Geoserver's, and make the CI use this new target. Or maybe more appropriate to bypass Geoserver requests tests the default test target? 

> I added specific comments on that topic. The current monitoring implementation is, I agree, not efficient, but the internal implementation could monitor paths once and notify multiple observers. In my opinion, that would be a better optimisation than piggybacking the Catalog implementation with Geoserver stuff.

I wasn't sure how to do that but now that you pointed it out, it really seems the obvious thing to do. I was focused making it work first. I'm pretty sure it will make some of the other corrections easier to implement.

fmigneault commented on cowbird/services/impl/geoserver.py:
```diff
@@ -18,30 +26,31 @@ class Geoserver(Service):
         """
         super(Geoserver, self).__init__(settings, name, **kwargs)
+        self.api_url = "{}/rest".format(self.url)
+        self.auth = (self.admin_user, self.admin_password)
```
These could be `None` based on initialization values. 
I'm not sure if requests using `auth` will like `(None, None)`, `("user", None)`, etc. 
Might need to set `None` directly if any is missing.

f-PLT replied:
I tested both, as well as (None, "password") and empty strings. Cowbird does not seem affected and the individual requests to Geoserver return the anticipated response, 401 : failed because lack of valid creds.

The structure did change since this comment, so it might not even be an issue anymore

fmigneault commented on AUTHORS.rst:
```diff
@@ -8,3 +8,4 @@ Contributors
 
 * Francis Charette Migneault <francis.charette-migneault@crim.ca> [`@fmigneault <https://github.com/fmigneault>`_]
 * David Byrns <david.byrns@crim.ca> [`@dbyrns <https://github.com/dbyrns>`_]
+* Francis Pelletier <francis.pelletier@crim.ca> [`@f-PLT <https://github.com/f-PLT>` _]
```
bad extra space after `>`

fmigneault commented on CHANGES.rst:
```diff
@@ -11,6 +11,8 @@ Features / Changes
 ~~~~~~~~~~~~~~~~~~~~~
 
 * Add an ssl verification setting.
+* Add Geoserver workspace and datastore creation/removal linked to user creation/removal
```
Missing end period.

fmigneault commented on config/config.example.yml:
```diff
@@ -25,8 +25,10 @@ services:
-    url: https://${HOSTNAME}/twitcher/ows/proxy/geoserver
+    url: http://${HOSTNAME}:8765/geoserver
     workspace_dir: ${WORKSPACE_DIR}
+    admin_user: admin
```
Also, makes sense if this corresponds to the hardcoded `/user_workspaces` in docker-compose file.

fmigneault commented on cowbird/services/impl/geoserver.py:
```diff
@@ -50,42 +59,335 @@ def permission_deleted(self, permission):
+        response_code = creation_response.status_code
+        if response_code == 201:
+            LOGGER.info("Datastore [%s] has been successfully created.", datastore_name)
+        elif response_code == 401:
+            LOGGER.error("The request has not been applied because it lacks valid authentication credentials.")
```
Does it always return 401 even when auth are provided but are incorrect? Not 403 instead?

f-PLT replied:
Yes, Geoserver always returns 401 for either missing or incorrect credentials

fmigneault replied:
Can you add a note about this because this is not the same behaviour across all API.

f-PLT replied:
Read up a bit on this, I guess it depends what you mean by incorrect. From my understanding, 403 is for when a user credentials are valid, but does not have permission to access a specific ressource (permanently forbidden), whereas 401 is for a lack or bad credentials (say, typo in the password).

In our use case, it should always be the Geoserver admin user, who has all the permissions necessary, so the only 403 we should get would be related to specific routes and their restrictions, like here : https://github.com/Ouranosinc/cowbird/blob/geoserver-setup/cowbird/services/impl/geoserver.py#L56

I did add a generic `else` to `geoserver_response_handling` decorator so we'll still catch the error codes with the operations that lead up to them for all the cases that fall outside of what I tested.

fmigneault replied:
Yes, your understanding of 401/403 is the correct one. 
I know some API frameworks incorrectly replace 403 by 401, or vice-versa in some cases. 
Naming conventions of "Unauthorized" vs "Forbidden" are also often swapped around erroneously.
This is why I was wondering if 403 was not needed also in this case.

fmigneault commented on cowbird/services/impl/geoserver.py:
```diff
@@ -50,42 +59,335 @@ def permission_deleted(self, permission):
+                                                                   datastore_name=datastore_name,
+                                                                   datastore_path=datastore_path)
+        response_code = configuration_response.status_code
+        if response_code == 200:
+            LOGGER.info("Datastore [%s] has been successfully configured.", datastore_name)
+        elif response_code == 401:
+            LOGGER.error("The request has not been applied because it lacks valid authentication credentials.")
+        elif response_code == 500:
+            LOGGER.error(configuration_response.text)
+        else:
+            LOGGER.error("There was an error configuring the following datastore: [%s]", datastore_name)
```
This recurring pattern would be worth using a decorator.

fmigneault commented on cowbird/services/impl/geoserver.py:
```diff
@@ -50,42 +59,335 @@ def permission_deleted(self, permission):
+        """
+        request_url = "{}/workspaces/".format(self.api_url)
+        payload = {"workspace": {"name": workspace_name, "isolated": "True"}}
+        request = requests.post(url=request_url, json=payload, auth=self.auth, headers=self.headers)
+        return request
```
All code instances like so are confusing because `request` variable will contain a `Response` object. 
They should be renamed to `response`.

fmigneault commented on cowbird/services/service.py:
```diff
@@ -12,11 +12,15 @@
 SERVICE_PRIORITY_PARAM = "priority"
 SERVICE_URL_PARAM = "url"
 SERVICE_WORKSPACE_DIR_PARAM = "workspace_dir"
+SERVICE_ADMIN_USER = "admin_user"
+SERVICE_ADMIN_PASSWORD = "admin_password"
```
Might need a `# nosec` comment if the security checks triggers on this.

fmigneault commented on tests/test_service_factory.py:
```diff
@@ -21,7 +21,7 @@ def setUpClass(cls):
             "services": {
                 "Catalog": {"active": False},
-                "Geoserver": {"active": True, "url": "", "workspace_dir": ""},
+                "Geoserver": {"active": True, "url": "", "workspace_dir": "", "admin_user": "", "admin_password": ""},
```
Could make use of the `SERVICE_<>` parameter names.

f-PLT commented:
> @f-PLT @dbyrns
> 
> > > The tests are failing right now at the Install dependencies (prior to this PR), looks like it's hanging on docformatter-0.5.6.tar.gz . I haven't investigated yet, in case it's a known problem.
> > 
> > 
> > I will let @fmigneault gives its input on this problem.
> 
> I will need an output log or something to analyze the issue.
> Package `docformatter` has not had any new version since Dec 2020, so I doubt it is (directly) the cause.

The output from the cancelled job, after 6 hours running (https://github.com/Ouranosinc/cowbird/runs/3509241127?check_suite_focus=true) : 

> INFO: pip is looking at multiple versions of docformatter to determine which version is compatible with other requirements. This could take a while.
> Collecting docformatter
>   Downloading docformatter-1.3.1.tar.gz (16 kB)
>   WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='files.pythonhosted.org', port=443): Read timed out. (read timeout=15)")': /packages/35/52/41c0152a44873c8f11f0f9fe21b680588ae0e2b20e4a0a9b4812f8decbdc/docformatter-1.3.tar.gz
>   Downloading docformatter-1.3.tar.gz (16 kB)
> ...
> INFO: pip is looking at multiple versions of docformatter to determine which version is compatible with other requirements. This could take a while.
> ...
>   WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='files.pythonhosted.org', port=443): Read timed out. (read timeout=15)")': /packages/d9/0d/c642dd005280fcf4ae1927e47bc838bab3aefdb6e3a4555105748535e584/docformatter-0.5.6.tar.gz
>   Downloading docformatter-0.5.6.tar.gz (12 kB)

Since I didn't add new dependencies, I'm guessing it's because one of our unlocked dependencies changed and it broke something? 

From the github actions page, it seems to have happened quite a few times, with the first recent occurrence being this one : https://github.com/Ouranosinc/cowbird/actions/runs/1137547725

f-PLT committed:
Fix dependency issue

Asteroid package version 2.7.3-py3-none-any.whl
was still installed, which cause issue with compatibility
of other packages.

fmigneault commented:
@f-PLT 
I tried full uninstall of `docformatter` on my side, followed by `make check-docf`. 
Package `docformatter==1.4` got installed and ran without problem. 
I have `pip 21.2.4` installed, which is the latest one available: https://pypi.org/project/pip/
I'm not able to reproduce the error you are seeing. Maybe setting `docformatter>=1.4` will help resolve it?

f-PLT committed:
Change request to response

f-PLT commented:
@fmigneault 

I found the problem!

I was able to reproduce the issue with a new conda environment.

It's actually  `astroid`. Even with the the `!=2.7`, it still found a way to install this version : `astroid-2.7.3-py3-none-any.whl` 

changing it to `astroid>=2.6,!=2.7.*,<3`  fixed it locally, will be pushing the change to see if it works server side

f-PLT commented:
Hmm, the change did not seem to take, as I can see the same weird version for `astroid` in the github actions report. I'll see if it sorts itself with futur commits. If not, I'll try to fix it to <2.7

fmigneault commented:
@f-PLT 
I strongly believe it is the new `pip` resolver that causes issue. I'm facing similar problem as of the last 1-2 days on ReadTheDocs failing because of long resolver memory/times use. 
By adding stricter dependencies, sometimes the set of possible versions gets resolved more easily so that it moves on to the next dependency quicker. Probably just happens to be `astroid` in this case. 
Normally, it should resolve the version fast though, because `astroid` is strongly coupled with `pylint`, which is set as `pylint>=2.7,<2.8` in this case. Maybe the pinned version of `astroid` mismatches `pylint`, so it force install 2.7 ?

fmigneault commented:
@f-PLT 
Looks to me like things are working as expected with last test run.
Error is related to one of my comment about the 'hardcoded password' in the tests:
https://github.com/Ouranosinc/cowbird/runs/3549388289
Just need to apply  `# nosec: B105` next to it.

f-PLT committed:
Correct typos and other small changes

f-PLT committed:
Refactor for response handling decorator

f-PLT committed:
Fix typo

f-PLT committed:
Move shapefile datastore monitoring to Geoserver class

f-PLT committed:
Cleanup of user_workspaces folder

f-PLT committed:
Change hardcoded workspace to env var

f-PLT committed:
Makefile now uses env.example for docker-compose

f-PLT committed:
Make geoserver requests tests optionnal

This way, the tests also won't be run by github

fmigneault commented:
@f-PLT 
Might be of interest: [`e27ca78` (#465)](https://github.com/Ouranosinc/Magpie/pull/465/commits/e27ca7858ab9f1c718419146e6478ba85950d0a0)

f-PLT committed:
Implement shapefile publishing as Task

f-PLT committed:
Update standalone geoserver tests

f-PLT commented:
> Might be of interest: [`e27ca78` (#465)](https://github.com/Ouranosinc/Magpie/pull/465/commits/e27ca7858ab9f1c718419146e6478ba85950d0a0)

Indeed, as it still looks like a problem. I'll look to implement it early next week.

f-PLT committed:
Implement pip resolver workaround

Current resolver leads to endless failures,
switching to legacy resolver.

f-PLT committed:
Docstring formatting

f-PLT committed:
Fix pycodestyle version

Attempt to fix error in github ci

f-PLT committed:
Fix check errors

f-PLT committed:
Add missing period in docker .env.example

f-PLT commented:
There's the Codacy failure I'm not sure how to handle, as we get a warning about storing passwords when in our case it's a key name for kwargs.

f-PLT committed:
Add comment to `geoserver_response_handling` decorator

fmigneault commented:
> There's the Codacy failure I'm not sure how to handle, as we get a warning about storing passwords when in our case it's a key name for kwargs.

@f-PLT 
Issue is now ignored.

fmigneault commented on Makefile:
```diff
@@ -594,21 +622,30 @@ test: test-all	## alias for 'test-all' target
 test-only:  ## run all tests, but without prior dependency check and installation
 	@echo "Running tests..."
-	@bash -c '$(CONDA_CMD) pytest tests -vv --junitxml "$(APP_ROOT)/tests/results.xml"'
+	@bash -c '$(CONDA_CMD) pytest tests -vv -m "not geoserver" --junitxml "$(APP_ROOT)/tests/results.xml"'
```
I think this `target` should remain "run all" regardless of applied tags.

f-PLT replied:
So leave this target to "run all" and create a new target that would then be the one called by Github actions?

fmigneault replied:
There could be separate job/matrix entries for api, cli, geoserver, etc.

f-PLT replied:
Would using `SPEC="not geoserver" test-custom` work in the context of Github actions? Could also be an alternative, especially for the python 3.9 test case if we don't want too much of an increase of checks in the report (would add 6 to cover the other markers).

Having separate entries for each marker in the main matrix could be helpful though, providing a quick view of what class of tests failed right off the bat.

fmigneault replied:
If you call specifically that spec often, a specific target is better. 
It is really based on usage convenience. 

One-of entry in the CI with `SPEC=...` is also valid, but you might need to update the test command since it expect only a target name: https://github.com/Ouranosinc/cowbird/blob/master/.github/workflows/tests.yml#L100-L101.

The only reason it is worth splitting by test markers in the test matrix is if some tests take specifically longer job-run time compared to others. If the full `test` target takes less or similar time to `checks`, I would just leave them all together. There is a limit on parallel jobs anyway (5 I think?), so over splitting won't make it faster. 

Other side note: 

This item: https://github.com/Ouranosinc/cowbird/blob/master/.github/workflows/tests.yml#L57
could be replaced by `[test-only]` since `install-dev` is already processed some steps before 
https://github.com/Ouranosinc/cowbird/blob/master/.github/workflows/tests.yml#L100-L101
(save up time of re-valiation of packages).

fmigneault reviewed (commented):
Code itself is fine. Only a few things to adjust regarding Makefile and requirements.

You can remove the `python_version < "3.6"` entries in requirements. 
We don't support lower python for cowbird anyway. 
Parts `; python_version >= "3.6"` can also be removed since this will by default always be the case.

f-PLT committed:
Remove python version checks from requirements

dbyrns commented on cowbird/services/impl/geoserver.py:
```diff
@@ -38,156 +81,174 @@ def __init__(self, settings, name, **kwargs):
+        if filename.endswith(".shp"):
+            workspace_name, shapefile_name = self._get_shapefile_info(filename)
+            # Small wait time to prevent unnecessary failure since shapefile is a multi file format
+            sleep(1)
```
Does this hang the whole monitoring thread for 10secs if I drop 10 .shp files?

dbyrns replied:
Maybe the validate_shapefile function could wait for 1 sec before doing its job? This way there is no sleep on the main thread.

f-PLT replied:
Good point, I'll move the sleep to function

dbyrns commented on cowbird/services/service.py:
```diff
@@ -12,15 +12,13 @@
 SERVICE_WORKSPACE_DIR_PARAM = "workspace_dir"
-SERVICE_ADMIN_USER = "admin_user"
-SERVICE_ADMIN_PASSWORD = "admin_password"
+SERVICE_ADMIN_USER = "admin_user"  # nosec: B105
```
Why those are not declared inside geoserver module as they are not required by any other service?

dbyrns commented on cowbird/services/impl/geoserver.py:
```diff
@@ -38,156 +81,174 @@ def __init__(self, settings, name, **kwargs):
 
+    # FSMonitor class functions
+    @staticmethod
+    def get_instance():
```
Why creating this function if you still use `ServiceFactory().get_service("Geoserver")` everywhere 😜

f-PLT replied:
Ooops haha! The Tasks where created before I implemented the dual inheritance with FSMonitor class and I must have implemented the abstract function mechanically from the Catalog implementation without thinking about it.

Looking at it now however, that function should return a monitor instance, and in both `geoserver.py` and `catalog.py`, we use it to return a Service instance. Is this really the behavior expected? If so I'd work on the description a bit to make it clearer.

dbyrns replied:
Sorry, I didn't realize it's a FSMonitor required function. The implementation is right though, that function is used here : https://github.com/Ouranosinc/cowbird/blob/master/cowbird/monitoring/monitor.py#L62 and should return an instance of `FSMonitor` which is really, in this case, a `Geoserver` instance.

fmigneault replied:
The pattern of `get_fsmonitor_instance` -> `get_instance` and similar feature for GeoServer is telling me `get_instance` should probably be an abstract method that classes must override with their own implementations.

fmigneault replied:
Oh, I mixed `FSMonitor` and `Monitor` classes. All good.

dbyrns commented on cowbird/services/impl/geoserver.py:
```diff
@@ -22,11 +27,49 @@
+        elif response_code == 500:
+            LOGGER.error("Operation [%s] failed : %s", operation, response.text)
+        else:
+            LOGGER.error("Operation [%s] failed with HTTP error code [%s]", operation, response_code)
```
I guess this is where a 404 not found (aka server is not running) would fall... We should throw something so that the autoretry can kick-in and send the request again in a couple of seconds. 
It seems that whatever happens with geoserver requests, we just swallow the error, log something and continue as if nothing happened. 
If the error is temporary, throwing it will cause the request to be retried. 
If the error is fatal, throw a recognizable exception so that geoserver's celery task can call `RequestTask.abort_chain` and avoid any downstream operations.

f-PLT replied:
Good find.

The other cases (401, 403 and most of the 500) are fatal errors not worth retrying as they are mostly `workspace/datastore already/doesn't exist` and retrying with the same parameters will only yield the same error.

f-PLT replied:
Ok, I was testing this and turns out 404 is also used by Geoserver when a ressource is not found (like trying to delete a workspace that does not exist), in which case retrying won't solve anything. I'll have to expand this decorator a bit to catch those cases and prevent useless long retries

dbyrns replied:
In fact, I was wrong with the 404, I thing that a timed-out request will yield a 504 Gateway Time-out

fmigneault replied:
Could also be a ReadTimeout 408 depending on how the client/server handle it.

f-PLT replied:
After some testing, what I see when the Geoserver is offline is the actual request failing (say `requests.get(...)` )  and raising an exception, so no error codes or anything gets returned. The task then auto-retries (because `RequestException` is in `RequestTask`'s auto-retry list) but it doesn't show up in the logs.

For the rest, I don't think there is a lot of "retry" potential, they're fatal errors that would hint at an errors in the code, duplicate users or  bad inputs.

I'll be wrapping the fatal errors in an Exception and raise it to stop the chain, and I'll work something up so we can get some feedback when Geoserver is unavailable.

dbyrns reviewed (changes requested):
Good job we're almost there. See my comments for some more requested changes.

f-PLT committed:
Replace ServiceFactory with get_instance

f-PLT committed:
Move credential keys to `geoserver.py`

f-PLT committed:
Add specificity to the get_instance function

f-PLT committed:
Move wait time to worker task

f-PLT committed:
Create test target specific for CI

Special target for CI execution that skips
Geoserver requests tests, as they require a live
Geoserver instance.

f-PLT committed:
Fix linting

f-PLT committed:
`geoserver_response_handling` now raises exceptions

f-PLT commented:
Same issue as before concerning Codacy, since the variables were moved.

dbyrns commented on cowbird/services/impl/geoserver.py:
```diff
@@ -479,31 +494,37 @@ def _remove_shapefile_request(self, workspace_name, datastore_name, filename):
+
+class GeoserverError(Exception):
+    """
+    Generic Geoserver error used to break request chains.
```
Maybe just say why : `...because not covered by the ``RequestTask`` retry_for exceptions`

dbyrns reviewed (approved):
LGTM

f-PLT committed:
Add details to GeoserverError docstring

f-PLT committed:
Small refactor of `geoserver_response_handling`

Streamlining of error messages.
`geoserver_response_handling` now only issues a log warning
if a workspace already exists, as it has no other consequence
and raising an exception in this case would then prevent
the datastore from being created

fmigneault commented on Makefile:
```diff
@@ -646,6 +646,11 @@ test-geoserver: install-dev install		## run Geoserver requests tests against a c
 	@echo "Running local tests..."
 	@bash -c '$(CONDA_CMD) pytest tests -vv -m "geoserver" --junitxml "$(APP_ROOT)/tests/results.xml"'
 
+.PHONY: test-github
+test-github:  ## test target used by github's ci, runs all tests except Geoserver requests, without prior dependency check and installation
+	@echo "Running tests..."
+	@bash -c '$(CONDA_CMD) pytest tests -vv -m "not geoserver" --junitxml "$(APP_ROOT)/tests/results.xml"'
```
I still don't understand the logic here. 
What is the point to skip geoserver tests? 
We want the test suite to validate that features still work each time we change something.

f-PLT replied:
`geoserver` might not be the best marker name for this, as it does not convey the online nature of this test suite. I had a chance to talk about this with @dbyrns and it was the solution we ended up with.

Those tests are only covering the specific requests to a Geoserver instance, hence they require a live Geoserver. I agree that this kind of tests  would be better suited for a notebook test at the birdhouse level like the other integration tests, but this was somewhat out of scope for this PR.

However, this series of tests were already written by that time and were quite useful while developing. Making Github run a Geoserver instance in it's workflow seemed a bit overkill (and would become redundant once the notebook tests are created) and transforming them to use a Mock reduces their effectiveness, as we wouldn't know if the requests are really valid from the tests. So I took the same approach as with Birdy that skips the `slow` and `online` tests in its Github workflow.

fmigneault replied:
Ok. I think a note about this specific reason (GeoServer instance online needed) would be important to add above the target, so we can remember about this after a few months. Maybe `online` marker would make this more explicit (the test could have both `geoserver` and `online`, and we skip `online` tests in CI which seems more natural).

I will leave it up to @dbyrns to take the final decision regarding if we should test something.
I still think that at least a minimal test should be added with a mock instance to ensure that the most basic operations work. We don't have to test all cases, just the most common workflow. I have found (with Weaver PR for example) that the full cycle of (1) develop component, (2) deploy in birdhouse, (3) test-notebooks, (4) oups! broken, (5) back to (1) is extremely tedious and time consuming. It's even worst when the broken thing is not even related to (1)'s feature but a side effect of it.

f-PLT replied:
I like the dual marker solution.

As for testing, right now the `test_request_task` class test our Geoserver service's user creation/removal, and file creation/removal implementations . I could use the same approach and patch the http requests so the CI tests reach all the way to the Geoserver requests. Less involved than mocking Geoserver, but combined with the `online` tests, it would give us a pretty good coverage at the component level.

It would only leave the shapefile publishing aspect untested, as it's squarely in the integration test category. Right now I test it manually (all the way to accessing it via WFS), but I think I could add it to the online tests with a bit of tinkering without it taking too long.

dbyrns replied:
I also like the dual marker solution. 
As for what should be tested, unlike Weaver, the Geoserver component is a small sub-system of Cowbird and should not evolve a lot. Mocking every services used by Cowbird would be very time consuming for the investment return. Given that we still have an online test that can validate the end-to-end functionnality, I think that's enough.

fmigneault replied:
Sounds fair. Make sure to apply `geoserver` to the relevant unit-tests that will run offline in the CI, this way we can run specifically those as needed.

fmigneault commented:
> Same issue as before concerning Codacy, since the variables were moved.

I updated the Codacy config. It should not trigger anymore. 
A separate sub-tool was flagging security issues, while the bandit tool was ignoring them correctly as per `nosec`.

f-PLT committed:
final touch up to tests

f-PLT committed:
Fix check-lint error
